File: Solutions/하루 한문제 코테준비/경주로건설_프로그래머스_67259.py
import copy
from collections import deque
dx,dy = [0,0,1,-1],[1,-1,0,0]
def bfs(sx,sy):
    q = deque()
    q.append((sx,sy,0,0))

    while q:
        x,y,direction,value = q.popleft()
        for i in range(4):
            nx,ny = x+dx[i],y+dy[i]
            if(0<=nx<n and 0<=ny<n):
                if(cpyBoard[nx][ny]==0):
                    if(x==0 and y==0):
                        newValue = value+100
                    else:
                        if(direction==i):
                            newValue = value+100
                        else:
                            newValue = value+600

                    if(cost[nx][ny]>=newValue):
                        cost[nx][ny] = newValue
                        q.append((nx,ny,i,newValue))




def solution(board):
    global result,n,cost,cpyBoard
    cpyBoard = copy.deepcopy(board)
    n = len(board)
    cost = [[int(1e9) for i in range(n)]for j in range(n)]
    answer = 0

    cost[0][0] = 0
    bfs(0,0)
    # for i in cost:
    #     print(i)
    return cost[-1][-1]

print(solution(	[[0, 0, 0, 0, 0, 0, 0, 1], [0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 1, 0, 0], [0, 0, 0, 0, 1, 0, 0, 0], [0, 0, 0, 1, 0, 0, 0, 1], [0, 0, 1, 0, 0, 0, 1, 0], [0, 1, 0, 0, 0, 1, 0, 0], [1, 0, 0, 0, 0, 0, 0, 0]]))

File: Solutions/하루 한문제 코테준비/javascript_solve/경주로건설_프로그래머스_67259.js
const bfs = (sx,sy,n,cpyBoard,cost)=>{
    const [dx,dy] = [[0,0,1,-1],[1,-1,0,0]];
    const q = [];
    q.push([sx,sy,0,0]);
    let newValue;
    while(q.length !== 0){
        const [x,y,direction,value] = q.shift();
        for(let i=0; i<4; i++){
            const [nx,ny] = [x+dx[i],y+dy[i]];
            // console.log(nx,ny,cost)
            if(0<=nx && nx<n && 0<=ny&& ny<n){
                if(cpyBoard[nx][ny]===0){
                    if(x===0 && y===0) newValue = value+100;
                    else{
                        if(direction===i) newValue = value+100;
                        else newValue = value+600;
                    }
                    if(cost[nx][ny]>=newValue){
                        cost[nx][ny] = newValue;
                        q.push([nx,ny,i,newValue]);
                    }
                }
                    
            }

        }
                
    }
    return cost[n-1][n-1];
}
function solution(board) {
    var cost;
    var answer = 0;
    var n = board.length;
    var cpyBoard = [...board];
    cost = Array.from(Array(board.length), ()=>Array(board.length).fill(1e9));
    cost[0][0] = 0;
    const result = bfs(0,0,n,cpyBoard,cost);

    return result;
}
console.log(solution([[0, 0, 0, 0, 0, 0, 0, 1], [0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 1, 0, 0], [0, 0, 0, 0, 1, 0, 0, 0], [0, 0, 0, 1, 0, 0, 0, 1], [0, 0, 1, 0, 0, 0, 1, 0], [0, 1, 0, 0, 0, 1, 0, 0], [1, 0, 0, 0, 0, 0, 0, 0]]));